infrastructure/network/addressmanager/addressmanager.go

```diff
@@ -5,1264 +5,254 @@
-	if knownAddress.subnetworkID == nil {
-		am.fullNodeTriedAddressBucketArray[bucketIndex] = append(am.fullNodeTriedAddressBucketArray[bucketIndex], knownAddress)
-		return
-	}
-
-	if _, ok := am.subnetworkTriedAddresBucketArrays[*knownAddress.subnetworkID]; !ok {
-		am.subnetworkTriedAddresBucketArrays[*knownAddress.subnetworkID] = &triedAddressBucketArray{}
-		for i := range am.subnetworkTriedAddresBucketArrays[*knownAddress.subnetworkID] {
-			am.subnetworkTriedAddresBucketArrays[*knownAddress.subnetworkID][i] = nil
-		}
-	}
-	am.subnetworkTriedAddresBucketArrays[*knownAddress.subnetworkID][bucketIndex] = append(am.subnetworkTriedAddresBucketArrays[*knownAddress.subnetworkID][bucketIndex], knownAddress)
+type netAddressWrapper struct {
+	netAddress *appmessage.NetAddress
+	isBanned   bool
+	isLocal    bool
 }
 
-// expireNew makes space in the new buckets by expiring the really bad entries.
-// If no bad entries are available we look at a few and remove the oldest.
-func (am *AddressManager) expireNew(subnetworkID *subnetworkid.SubnetworkID, bucketIndex int) {
-	// First see if there are any entries that are so bad we can just throw
-	// them away. otherwise we throw away the oldest entry in the cache.
-	// We keep track of oldest in the initial traversal and use that
-	// information instead.
-	var oldest *KnownAddress
-	newAddressBucketArray := am.newAddressBucketArray(subnetworkID)
-	for addressKey, knownAddress := range newAddressBucketArray[bucketIndex] {
-		if knownAddress.isBad() {
-			log.Tracef("expiring bad address %s", addressKey)
-			delete(newAddressBucketArray[bucketIndex], addressKey)
-			knownAddress.referenceCount--
-			if knownAddress.referenceCount == 0 {
-				am.decrementNewAddressCount(subnetworkID)
-				delete(am.addressIndex, addressKey)
-			}
-			continue
-		}
-		if oldest == nil {
-			oldest = knownAddress
-		} else if !knownAddress.netAddress.Timestamp.After(oldest.netAddress.Timestamp) {
-			oldest = knownAddress
-		}
-	}
+// AddressKey represents a "string" key of the ip addresses
+// for use as keys in maps.
+type AddressKey string
 
-	if oldest != nil {
-		addressKey := NetAddressKey(oldest.netAddress)
-		log.Tracef("expiring oldest address %s", addressKey)
+// ErrAddressNotFound is an error returned from some functions when a
+// given address is not found in the address manager
+var ErrAddressNotFound = errors.New("address not found")
 
-		delete(newAddressBucketArray[bucketIndex], addressKey)
-		oldest.referenceCount--
-		if oldest.referenceCount == 0 {
-			am.decrementNewAddressCount(subnetworkID)
-			delete(am.addressIndex, addressKey)
-		}
-	}
+// NetAddressKey returns a key of the ip address to use it in maps.
+func netAddressKey(netAddress *appmessage.NetAddress) AddressKey {
+	key := make([]byte, len(netAddress.IP), len(netAddress.IP)+2)
+	copy(key, netAddress.IP)
+	return AddressKey(append(key, byte(netAddress.Port), byte(netAddress.Port>>8)))
```

Comment:
I don't like the bit-magic happening with the port number.
Better use `binary.LittleEndian.PutUint16` instead.